exp/controllers/machinepool_controller_phases.go

```diff
@@ -60,11 +60,21 @@ func (r *MachinePoolReconciler) reconcilePhase(mp *expv1.MachinePool) {
 		mp.Status.SetTypedPhase(expv1.MachinePoolPhaseProvisioned)
 	}
 
-	// Set the phase to "running" if there is a NodeRef field.
-	if mp.Status.InfrastructureReady && len(mp.Status.NodeRefs) == int(mp.Status.ReadyReplicas) {
+	// Set the phase to "running" if the number of ready replicas is equal to desired replicas.
+	if mp.Status.InfrastructureReady && *mp.Spec.Replicas == mp.Status.ReadyReplicas {
```

Comment:
changed this to match MachineDeployment: MachinePoolPhaseProvisioned means infra ready, MachinePoolPhaseRunning means infra ready + all replicas ready

Reply:
Only loosely related to your changes, but this function and the collection of various if statements reads to me like it's possible that the phase could be more than one thing. The phase will be set to the last conditional that evaluates as true, which seems misleading and error prone.

If conditions are mutually exclusive and the phase is singular, I would expect there to be only one set within the function. If that holds, then perhaps, consider using a switch statement to better indicate mutual exclusivity of the states.

Reply:
I actually asked myself the exact same question when I was first looking at this phase logic (and the machine deployment + kcp one which is the same). I think it was actually done that way intentionally and ordered by precedence. For example, it can be both provisioned and ready (ready takes precedence), it could be both ready and deleting (deleting takes precedence), and it can be both ready and scaling up (scaling up takes precedence). So I added scaling up and scaling down after ready but before deleting intentionally (same order used in KCP and MD).

But now that you bring it up and I think about it again I do think that quite fragile and a bug could easily be introduced. Maybe it'd be better to inverse the order and use a switch statement. Although with a switch the order of the cases would still matter (ie. Deleting would need to be checked before Ready for example) so I don't know if it's much of an improvement, would just be more readable 🤷‍♀️

Reply:
If there is only one state (phase) the resource can be in given the current state of the resource, having multiple sets with dependent order feels really awkward.

Reply:
I agree. Is it okay with you if I open a separate PR to re-organize this + MD + KCP phases? I'd rather not block this one on a bigger code cleanup.

Reply:
I believe the long term goal is to remove "Phases" and standardize on conditions in the future, might not be worth putting too much effort into